### PR TITLE
Add a `--region` option to R2's bucket command

### PR DESCRIPTION
**Pull request by sdnts**
What this PR solves / how to test:
We're planning on introducing region hints for bucket creation, so this lets users configure it from Wrangler

Associated docs issues/PR:

- [insert associated docs issue(s)/PR(s)]

Author has included the following, where applicable:

- [x] Tests
- [x] Changeset

Reviewer has performed the following, where applicable:

- [x] Checked for inclusion of relevant tests
- [x] Checked for inclusion of a relevant changeset
- [x] Checked for creation of associated docs updates
- [ ] Manually pulled down the changes and spot-tested

Fixes # [insert issue number].

### CODE REVIEW

File: packages/wrangler/src/__tests__/r2.test.ts [hunk 1]
```diff
@@ -106,7 +106,10 @@ describe("r2", () => {
 			  -c, --config   Path to .toml configuration file  [string]
 			  -e, --env      Environment to use for operations and .env files  [string]
 			  -h, --help     Show help  [boolean]
-			  -v, --version  Show version number  [boolean]"
+			  -v, --version  Show version number  [boolean]
+
+			Options:
```
Comment (sdnts) [inline packages/wrangler/src/__tests__/r2.test.ts:111]: So what is the difference between an `Option` and a `Flag`? Do we prefer one over the other? I saw that R2's command used options, so I stuck to that

File: packages/wrangler/src/__tests__/r2.test.ts [hunk 2]
```diff
@@ -106,7 +106,10 @@ describe("r2", () => {
...
+			  -r, --region  A hint to control where the bucket will be created  [string] [choices: \\"auto\\", \\"WNAM\\", \\"ENAM\\", \\"WEUR\\", \\"APAC\\", \\"OC\\", \\"EEUR\\", \\"SAM\\", \\"AFR\\", \\"ME\\", \\"CHINA\\"]"
```
Comment (sdnts) [inline packages/wrangler/src/__tests__/r2.test.ts:112]: Who would I get to review this copy? A technical writer, or are y'all responsible?

File: packages/wrangler/src/__tests__/r2.test.ts [hunk 3]
```diff
@@ -278,10 +340,8 @@ describe("r2", () => {
 					`r2 object put bucketName-object-test/wormhole-img.png --file ./wormhole-img.png`
 				)
 			).rejects.toThrowErrorMatchingInlineSnapshot(`
-			"Error: Wrangler only supports uploading files up to ${prettyBytes(
-				MAX_UPLOAD_SIZE
-			)} in size
-			wormhole-img.png is ${prettyBytes(TOO_BIG_FILE_SIZE)} in size"
+			"Error: Wrangler only supports uploading files up to 315 MB in size
+			wormhole-img.png is 316 MB in size"
```
Comment (sdnts) [inline packages/wrangler/src/__tests__/r2.test.ts:344]: Somehow this snapshot was out of date, so I updated it

File: packages/wrangler/src/r2/helpers.ts [hunk 4]
```diff
@@ -31,11 +31,12 @@ export async function listR2Buckets(
...
 export async function createR2Bucket(
 	accountId: string,
-	bucketName: string
+	bucketName: string,
+	regionHint?: string
 ): Promise<void> {
 	return await fetchResult<void>(`/accounts/${accountId}/r2/buckets`, {
 		method: "POST",
-		body: JSON.stringify({ name: bucketName }),
+		body: JSON.stringify({ name: bucketName, regionHint }),
```
Comment (sdnts) [inline packages/wrangler/src/r2/helpers.ts:39]: The actual implementation isn't released yet, but I'll make sure to hold this PR till it is.

File: packages/wrangler/src/__tests__/r2.test.ts (hunk 3 above)
Comment (rozenmd) [reply packages/wrangler/src/__tests__/r2.test.ts:344]: To update it, did you run `npm run test -- -u`?

File: packages/wrangler/src/__tests__/r2.test.ts [hunk 5]
```diff
@@ -143,6 +149,40 @@ describe("r2", () => {
 			          `);
 			});
 
+			it("should error for unknown regions", async () => {
```
[the remark this answers is not in the document]
Comment (rozenmd) [reply packages/wrangler/src/__tests__/r2.test.ts:152]: Yeah no need to test our libs (yargs already has tests for choices)

**Comment by isaac-mcfadyen** [comment]
Small piece of feedback (Region support looks awesome!)

In the help text the region names are currently acronyms and some of them won't be known by people, especially since they seem Cloudflare-specific or not commonly used (searching WNAM the first result is a Cloudflare page).

I wonder if we could put the full, expanded name in brackets after it (in the help message) so people don't have to search to find them all? Something like:
```
APAC (Asia-Pacific)
WNAM (Western North America)
```

**Comment by sdnts** [comment]
Great callout! Let me see if I can get yargs to show this somehow

File: packages/wrangler/src/__tests__/r2.test.ts (hunk 3 above)
Comment (sdnts) [reply packages/wrangler/src/__tests__/r2.test.ts:344]: Yep!

File: packages/wrangler/src/__tests__/r2.test.ts (hunk 2 above)
Comment (rozenmd) [reply packages/wrangler/src/__tests__/r2.test.ts:112]: Normally the product's PM chimes in with their thoughts

File: packages/wrangler/src/r2/index.ts [hunk 6]
```diff
@@ -221,22 +221,30 @@ export function r2(r2Yargs: CommonYargsArgv) {
...
+							describe: "The name of the new bucket",
+							type: "string",
+							demandOption: true,
+						})
+						.option("location", {
+							describe: "A hint to control where the bucket will be created",
+							alias: "l",
+							type: "string",
+							default: "auto",
+							choices: ["auto", "apac", "eeur", "enam", "weur", "wnam"],
```
Comment (sdnts) [inline packages/wrangler/src/r2/index.ts:235]: So as @isaac-mcfadyen suggested, I tried to explain what these abbreviations mean. I tried a couple of things:

1. I changed the `choices` to include the meanings of locations, and coupled that with the `coerce` option, but that forces the user to type in `apac (Asia-Pacific)` in its entirety, which is of course not ideal
2. I tried adding `( apac = Asia-Pacific, ... )` to the `describe` string above, which works but I don't like it very much.
3. Linking to https://developers.cloudflare.com/r2/buckets/data-location/#available-hints is also not ideal.

I can go ahead with `2.` unless people here have better ideas?

Comment (rozenmd) [reply packages/wrangler/src/r2/index.ts:235]: Check this out - facing similar issues for D1 @sdnts: https://github.com/cloudflare/workers-sdk/pull/2997

Comment (sdnts) [reply packages/wrangler/src/r2/index.ts:235]: Yeah I think I'm going to commit 2. We should be consistent with what locations we show though: I see your PR expands "weur" to "Europe (West)". I'm going off of https://developers.cloudflare.com/r2/buckets/data-location/#location-hints

**Pull request closed without merging**